x-pack/plugins/security_solution/public/detections/components/alerts_table/timeline_actions/use_add_to_case_actions.tsx

```diff
@@ -5,68 +5,121 @@
+  const { cases: casesUi } = useKibana().services;
   const casePermissions = useGetUserCasesPermissions();
-  const insertTimelineHook = useInsertTimeline;
+  const casesToasts = casesUi.hooks.useCasesToast();
+  const hasWritePermissions = casePermissions?.crud ?? false;
 
-  const addToCaseActionProps = useMemo(
-    () =>
-      ecsData?._id
-        ? {
-            ariaLabel,
-            event: { data: nonEcsData ?? [], ecs: ecsData, _id: ecsData?._id },
-            useInsertTimeline: insertTimelineHook,
-            casePermissions,
-            appId: APP_UI_ID,
+  const caseAttachments: CaseAttachments = useMemo(() => {
+    return ecsData?._id
+      ? [
+          {
+            alertId: ecsData?._id ?? '',
+            index: ecsData?._index ?? '',
             owner: APP_ID,
-            onClose: afterCaseSelection,
-          }
-        : null,
-    [ecsData, ariaLabel, nonEcsData, insertTimelineHook, casePermissions, afterCaseSelection]
-  );
-  const hasWritePermissions = casePermissions?.crud ?? false;
-  const addToCaseActionItems = useMemo(
-    () =>
+            type: CommentType.alert,
+            rule: casesUi.helpers.getRuleIdFromEvent({ ecs: ecsData, data: nonEcsData ?? [] }),
+          },
+        ]
+      : [];
+  }, [casesUi.helpers, ecsData, nonEcsData]);
+
+  const createCaseFlyout = casesUi.hooks.getUseCasesAddToNewCaseFlyout({
+    attachments: caseAttachments,
+    onSuccess: async (theCase: Case) => {
+      casesToasts.showSuccessAttach(theCase);
+    },
+    onClose: onMenuItemClick,
+  });
+
+  const selectCaseModal = casesUi.hooks.getUseCasesAddToExistingCaseModal({
+    attachments: caseAttachments,
+    onRowClick: (theCase?: Case) => {
+      if (theCase !== undefined) {
+        casesToasts.showSuccessAttach(theCase);
+      }
+    },
+    onClose: onMenuItemClick,
+  });
+
+  const handleAddToNewCaseClick = useCallback(() => {
+    // TODO rename this, this is really `closePopover()`
+    onMenuItemClick();
+    createCaseFlyout.open();
+  }, [onMenuItemClick, createCaseFlyout]);
+
+  const handleAddToExistingCaseClick = useCallback(() => {
+    // TODO rename this, this is really `closePopover()`
+    onMenuItemClick();
+    selectCaseModal.open();
+  }, [onMenuItemClick, selectCaseModal]);
+
+  const addToCaseActionItems = useMemo(() => {
+    if (
       [
         TimelineId.detectionsPage,
         TimelineId.detectionsRulesDetailsPage,
         TimelineId.active,
       ].includes(timelineId as TimelineId) &&
-      hasWritePermissions &&
-      addToCaseActionProps
-        ? [
-            timelinesUi.getAddToExistingCaseButton(addToCaseActionProps),
-            timelinesUi.getAddToNewCaseButton(addToCaseActionProps),
-          ]
-        : [],
-    [addToCaseActionProps, hasWritePermissions, timelineId, timelinesUi]
-  );
+      hasWritePermissions
+    ) {
+      return [
+        // add to existing case menu item
+        <EuiContextMenuItem
```

Comment:
imo the markup for this should live in cases, and this change is more along the lines of timelinesUi.getAddToExistingCaseButton(addToCaseActionProps) -> cases.getAddToExistingCaseButton(addToCaseActionProps)

Reply:
@kqualters-elastic We decided against this because this is logic that belongs to the plugin consumer and not to cases.

Cases provides the API to open the flyout and the modal but should not decide how the plugins expose the UI to open it. 

If we provide the `EuiContextMenuItem` directly, then that's all plugins are left to use. What if in the future plugins want a button? a link? or a different UI element that doesn't support a `EuiContextMenuItem`? We'll have to then start providing wrappers for every possible button.

Reply:
In that case you could simply add a type or style prop to the 1 exported action component, an enum that is something like 'contextMenu' | 'icon' | 'text'. that way the business logic for when/how to render each of the case flow components lives in one and only one place.

Reply:
I agree with both @academo and @cnasikas. There will very likely be a multitude of different UI elements used to add attachments to cases. 

If we see ourselves commonly reimplementing a `EUIContextMenuItem` that renders "Add to case", then I think it'd make sense for the cases plugin to expose an `AddAttachmentToCaseEUIContextMenuItem` (name debatable) React Component. However, I don't think that we should do so at this point in time. Let's not prematurely generalize a solution here.

Reply:
As it is now, in this pr, there is about ~100 lines of duplicated code in this file and the code in the action item definitions in x-pack/plugins/observability/public/pages/alerts/containers/alerts_table_t_grid/alerts_table_t_grid.tsx . 100 lines x 2 is maintainable no doubt, however if this expands to include upcoming features like showing associated cases to an alert (https://github.com/elastic/security-team/issues/2209) and logic to link back to each, and possibly all of this logic being replicated a third time in the upcoming stack alerts table, 200 lines of duplicated code for display related business logic can quickly balloon into 1k+. Compare that with the <20 duplicated lines across plugin consumers when done with components that are connected to a shared redux store (or a single global plugin context, that's all redux is) like in this previous pr https://github.com/elastic/kibana/pull/114360/files . 

Additionally this approach is notoriously brittle when refactoring. Recently I added a component that not obviously needed to have the cases context provider in it's component tree https://github.com/elastic/kibana/pull/126195/files to not blow up. x-pack/plugins/security_solution/public/timelines/components/side_panel/event_details/index.tsx is a component that renders 3 components in an if/else based on a boolean flag: ExpandableEventTitle, ExpandableEvent and EventDetailsFooter along with some styling depending on where in the app it's being used, either in the alert table or timeline flyout. It is not obvious whatsoever to a developer that EventDetailsFooter will render just fine when said boolean is true, because the context provider is already present as a parent where that flavor of the component is used, but throw an uncaught exception when the boolean is false unless the <CasesContext /> is added.

Reply:
Also isn't customizing buttons that do the exactly the same thing sort of antithetical to the stated goals of RAC and really sort of bad UX as it will destroy any user affordances that could be created by having all of the markup in 1 place? Buttons that do exactly the same thing but look or are labeled differently from one another is not a great UX, and building that capability in before having a valid reason to do so seems like a generalized solution to a problem that likely won't ever exist. Solution teams won't want to maintain 200 lines of components, hooks and css, nor will the Eui team want to update 10 different button implementations if needed when only 1 is possible.